### PR TITLE
Implement support for PostgreSQL 12 "standby.signal".

### LINKED ISSUE #44: Build error against PostgreSQL 12
Hi,

pg_auto_failover cannot be built against PostgreSQL 12:

```
+ PG_CONFIG=/usr/pgsql-12/bin/pg_config
+ /usr/bin/make -j3
ipaddr.c: In function 'findHostnameFromLocalIpAddress':
ipaddr.c:548:12: warning: unused variable 'address_len' [-Wunused-variable]
  socklen_t address_len;
            ^~~~~~~~~~~
ipaddr.c:547:26: warning: unused variable 'address' [-Wunused-variable]
  struct sockaddr_storage address;
                          ^~~~~~~
health_check_worker.c: In function 'HealthCheckWorkerLauncherMain':
health_check_worker.c:233:20: error: implicit declaration of function 'AllocSetContextCreateExtended'; did you mean 'AllocSetContextCreateInternal'? [-Werror=implicit-function-declaration]
  launcherContext = AllocSetContextCreateExtended(CurrentMemoryContext,
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                    AllocSetContextCreateInternal
health_check_worker.c:233:18: error: assignment to 'MemoryContext' {aka 'struct MemoryContextData *'} from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  launcherContext = AllocSetContextCreateExtended(CurrentMemoryContext,
                  ^
health_check_worker.c: In function 'BuildDatabaseList':
health_check_worker.c:363:9: error: implicit declaration of function 'heap_beginscan_catalog'; did you mean 'table_beginscan_catalog'? [-Werror=implicit-function-declaration]
  scan = heap_beginscan_catalog(pgDatabaseRelation, 0, NULL);
         ^~~~~~~~~~~~~~~~~~~~~~
         table_beginscan_catalog
health_check_worker.c:363:7: error: assignment to 'HeapScanDesc' {aka 'struct HeapScanDescData *'} from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  scan = heap_beginscan_catalog(pgDatabaseRelation, 0, NULL);
       ^
In file included from /usr/pgsql-12/include/server/postgres.h:46,
                 from health_check_worker.c:13:
health_check_worker.c:365:49: error: passing argument 1 of 'heap_getnext' from incompatible pointer type [-Werror=incompatible-pointer-types]
  while (HeapTupleIsValid(dbTuple = heap_getnext(scan, ForwardScanDirection)))
                                                 ^~~~
/usr/pgsql-12/include/server/c.h:626:48: note: in definition of macro 'PointerIsValid'
 #define PointerIsValid(pointer) ((const void*)(pointer) != NULL)
                                                ^~~~~~~
health_check_worker.c:365:9: note: in expansion of macro 'HeapTupleIsValid'
  while (HeapTupleIsValid(dbTuple = heap_getnext(scan, ForwardScanDirection)))
         ^~~~~~~~~~~~~~~~
In file included from health_check_worker.c:21:
/usr/pgsql-12/include/server/access/heapam.h:120:45: note: expected 'TableScanDesc' {aka 'struct TableScanDescData *'} but argument is of type 'HeapScanDesc' {aka 'struct HeapScanDescData *'}
 extern HeapTuple heap_getnext(TableScanDesc scan, ScanDirection direction);
                               ~~~~~~~~~~~~~~^~~~
cli_config.c: In function 'cli_config_check':
cli_config.c:82:6: warning: unused variable 'errors' [-Wunused-variable]
  int errors = 0;
      ^~~~~~
health_check_worker.c:378:19: error: implicit declaration of function 'HeapTupleGetOid'; did you mean 'HeapTupleIsValid'? [-Werror=implicit-function-declaration]
    entry->dboid = HeapTupleGetOid(dbTuple);
                   ^~~~~~~~~~~~~~~
                   HeapTupleIsValid
health_check_worker.c:387:15: error: passing argument 1 of 'heap_endscan' from incompatible pointer type [-Werror=incompatible-pointer-types]
  heap_endscan(scan);
               ^~~~
In file included from health_check_worker.c:21:
/usr/pgsql-12/include/server/access/heapam.h:119:40: note: expected 'TableScanDesc' {aka 'struct TableScanDescData *'} but argument is of type 'HeapScanDesc' {aka 'struct HeapScanDescData *'}
 extern void heap_endscan(TableScanDesc scan);
                          ~~~~~~~~~~~~~~^~~~
health_check_worker.c: In function 'HealthCheckWorkerMain':
health_check_worker.c:452:21: error: assignment to 'MemoryContext' {aka 'struct MemoryContextData *'} from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  healthCheckContext = AllocSetContextCreateExtended(CurrentMemoryContext,
                     ^
cli_service.c: In function 'cli_monitor_run':
cli_service.c:161:16: warning: unused variable 'existingPgSetup' [-Wunused-variable]
  PostgresSetup existingPgSetup = { 0 };
                ^~~~~~~~~~~~~~~
cli_service.c:155:7: warning: unused variable 'configFilePath' [-Wunused-variable]
  char configFilePath[MAXPGPATH];
       ^~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [<builtin>: health_check_worker.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:16: monitor] Error 2
make[1]: *** Waiting for unfinished jobs....
cli_show.c: In function 'cli_show_uri':
cli_show.c:309:10: warning: unused variable 'monitor' [-Wunused-variable]
  Monitor monitor = { 0 };
          ^~~~~~~
cli_create_drop_node.c: In function 'cli_create_monitor':
cli_create_drop_node.c:427:7: warning: unused variable 'ipAddr' [-Wunused-variable]
  char ipAddr[BUFSIZE];
       ^~~~~~
cli_create_drop_node.c:425:16: warning: unused variable 'pgSetup' [-Wunused-variable]
  PostgresSetup pgSetup = config.pgSetup;
                ^~~~~~~
cli_root.c: In function 'root_options':
cli_root.c:114:6: warning: unused variable 'debugCount' [-Wunused-variable]
  int debugCount = 0;
      ^~~~~~~~~~
pgsetup.c: In function 'get_pgpid':
pgsetup.c:403:7: warning: unused variable 'line' [-Wunused-variable]
  char line[BUFSIZE];
       ^~~~
pgsetup.c:402:6: warning: unused variable 'lineno' [-Wunused-variable]
  int lineno;
      ^~~~~~
cli_do_root.c: In function 'keeper_cli_keeper_setup_getopts':
cli_do_root.c:231:23: warning: unused variable 'errors' [-Wunused-variable]
  int c, option_index, errors = 0;
                       ^~~~~~
cli_do_root.c:231:9: warning: unused variable 'option_index' [-Wunused-variable]
  int c, option_index, errors = 0;
         ^~~~~~~~~~~~
cli_do_root.c:231:6: warning: unused variable 'c' [-Wunused-variable]
  int c, option_index, errors = 0;
      ^
monitor_config.c: In function 'monitor_config_set_setting':
monitor_config.c:442:13: warning: unused variable 'option' [-Wunused-variable]
  IniOption *option;
             ^~~~~~
monitor_pg_init.c: In function 'monitor_pg_init':
monitor_pg_init.c:60:7: warning: unused variable 'postgresInstanceExists' [-Wunused-variable]
  bool postgresInstanceExists = pg_setup_pgdata_exists(&pgSetup);
       ^~~~~~~~~~~~~~~~~~~~~~
```

Can you please take a look?

Thanks, Devrim

### PR DESCRIPTION
**Pull request by DimCitus**
In Postgres 12 the setup to start as a standby has changed. The standby
configuration parameters go in the main postgresql.conf file, and the
decision to start in standby mode depends on the existence of a new file,
named "standby.signal".

Adapting the code to this new system requires some refactoring.

Fixes #44.

### CODE REVIEW

File: src/bin/pg_autoctl/pgctl.c [hunk 1]
```diff
@@ -835,38 +907,78 @@ pg_ctl_promote(const char *pg_ctl, const char *pgdata)
...
+pg_setup_standby_mode(uint32_t pg_control_version,
+					  const char *configFilePath,
+					  const char *pgdata,
+					  ReplicationSource *replicationSource)
+{
+	NodeAddress *primaryNode = &(replicationSource->primaryNode);
+	char primaryConnInfo[MAXCONNINFO] = { 0 };
+
+	/* we ignore the length returned by prepare_primary_conninfo... */
+	int n = prepare_primary_conninfo(primaryConnInfo,
```
Comment (mtuncer) [inline src/bin/pg_autoctl/pgctl.c]: unused variable warning

Comment (DimCitus) [reply src/bin/pg_autoctl/pgctl.c]: Fixed now, thanks!

File: src/bin/pg_autoctl/pgctl.c [hunk 2]
```diff
@@ -835,38 +907,78 @@ pg_ctl_promote(const char *pg_ctl, const char *pgdata)
...
+
+	/* we ignore the length returned by prepare_primary_conninfo... */
+	(void) prepare_primary_conninfo(primaryConnInfo,
+									MAXCONNINFO,
+									primaryNode->host,
+									primaryNode->port,
+									replicationSource->userName,
+									replicationSource->password);
+
+	if (pg_control_version < 1200)
```
Comment (mtuncer) [inline src/bin/pg_autoctl/pgctl.c]: I did not know this, neat

File: src/bin/pg_autoctl/pgctl.c [hunk 3]
```diff
@@ -918,6 +1030,100 @@ escape_recovery_conf_string(char *destination, const char *recoveryConfString)
...
+static int
+prepare_primary_conninfo(char *primaryConnInfo, int primaryConnInfoSize,
+						 const char *primaryHost, int primaryPort,
+						 const char *replicationUsername,
+						 const char *replicationPassword)
+{
+	char primaryConnInfoRaw[MAXCONNINFO] = { 0 };
+	char *connInfoEnd = primaryConnInfoRaw;
+
+ 	connInfoEnd += make_conninfo_field_str(connInfoEnd, "host", primaryHost);
```
Comment (mtuncer) [inline src/bin/pg_autoctl/pgctl.c]: please consider using StringInfo if possible. using hardcoded lengths and unchecked accesses always worry me.

File: src/bin/pg_autoctl/pgctl.c [hunk 4]
```diff
@@ -918,6 +1030,100 @@ escape_recovery_conf_string(char *destination, const char *recoveryConfString)
...
+	connInfoEnd +=
+		make_conninfo_field_str(connInfoEnd, "user", replicationUsername);
+
+	if (replicationPassword != NULL)
+	{
+		connInfoEnd +=
+			make_conninfo_field_str(connInfoEnd, "password", replicationPassword);
+	}
+
+	return escape_recovery_conf_string(primaryConnInfoRaw, primaryConnInfo);
```
Comment (mtuncer) [inline src/bin/pg_autoctl/pgctl.c]: we keep appending to `primaryConnInfoRaw` by using connInfoEnd. At this line we completely override it value coming from `primaryConnInfo`. Could you check ?

Comment (mtuncer) [reply src/bin/pg_autoctl/pgctl.c]: you probably only need to reverse parameters here

File: src/bin/pg_autoctl/pgctl.c [hunk 5]
```diff
@@ -918,6 +1030,100 @@ escape_recovery_conf_string(char *destination, const char *recoveryConfString)
...
+	{
+		return false;
+	}
+
+	if (!pg_include_config(standbyConfigFilePath,
+						   AUTOCTL_SB_CONF_INCLUDE_LINE,
+						   AUTOCTL_SB_CONF_INCLUDE_REGEX,
+						   AUTOCTL_CONF_INCLUDE_COMMENT))
+	{
+		log_error("Failed to prepare \"%s\" with standby settings",
```
Comment (mtuncer) [inline src/bin/pg_autoctl/pgctl.c]: we have the signal file in place, but we can not include recovery config in postgresql.conf file. Is it okay to leave signal file.

File: src/bin/pg_autoctl/pgctl.c [hunk 6]
```diff
@@ -272,23 +313,19 @@ ensure_default_settings_file_exists(const char *configFilePath,
 	for (settingIndex = 0; settings[settingIndex].name != NULL; settingIndex++)
 	{
 		GUC *setting = &settings[settingIndex];
+		char *value = get_current_GUC_value(settings, setting->name, pgSetup);
```
Comment (mtuncer) [inline src/bin/pg_autoctl/pgctl.c]: we already know the setting that we got from settings collections. Passing the collection to for another search inside get_current_GUC_value looked unnecessary.
I think we can use
```c
get_current_GUC_value(setting, pgSetup)
```

File: src/bin/pg_autoctl/pgctl.c [hunk 7]
```diff
@@ -835,38 +907,78 @@ pg_ctl_promote(const char *pg_ctl, const char *pgdata)
...
+	log_trace("pg_write_recovery_conf");
 
 	/* build the contents of recovery.conf */
 	recoveryConfEnd += sprintf(recoveryConfEnd, "standby_mode = 'on'");
-	recoveryConfEnd += sprintf(recoveryConfEnd, "\nprimary_conninfo = ");
-	recoveryConfEnd += escape_recovery_conf_string(recoveryConfEnd, primaryConnInfo);
-	recoveryConfEnd += sprintf(recoveryConfEnd, "\nprimary_slot_name = '%s'",
-							   replicationSlotName);
-	recoveryConfEnd += sprintf(recoveryConfEnd, "\nrecovery_target_timeline = 'latest'");
+	recoveryConfEnd += sprintf(recoveryConfEnd,
```
Comment (mtuncer) [inline src/bin/pg_autoctl/pgctl.c]: could you consider using `StringInfo`, `appendStringInfo()` instead ? I know it was like this before. It is a good opportunity to fix that.

File: src/bin/pg_autoctl/pgctl.c (hunk 6 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/pgctl.c]: Fixed now, thanks!

File: src/bin/pg_autoctl/pgctl.c (hunk 7 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/pgctl.c]: See https://github.com/citusdata/pg_auto_failover/issues/54

File: src/bin/pg_autoctl/pgctl.c (hunk 5 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/pgctl.c]: Added comments about why I think we should do things that way.

File: src/bin/pg_autoctl/pgctl.c [hunk 8]
```diff
@@ -343,6 +405,29 @@ ensure_default_settings_file_exists(const char *configFilePath,
...
 
+/*
+ * get_current_GUC_value returns the current GUC value as in the settings
+ * parameter; replacing it with values found in pgSetup for the Postgres
+ * parameters "listen_addresses" and "port", allowing those to be dynamic.
+ */
+static const char *
+get_current_GUC_value(const char *name, const char *value, PostgresSetup *pgSetup)
+{
+	int settingIndex = 0;
```
Comment (mtuncer) [inline src/bin/pg_autoctl/pgctl.c]: this is not used anymore

**Pull request merged**